src/types.rs

```diff
@@ -136,3 +136,136 @@ impl LintPass for UnitCmp {
+        if let ExprCast(ref ex, _) = expr.node {
+            let (cast_from, cast_to) = (cx.tcx.expr_ty(&*ex), cx.tcx.expr_ty(expr));
+            if cast_from.is_numeric() && !in_external_macro(cx, expr.span) {
+                match (cast_from.is_integral(), cast_to.is_integral()) {
+                    (true, false)  => {
+                        match (&cast_from.sty, &cast_to.sty) {
+                            (&ty::TypeVariants::TyInt(i), &ty::TypeVariants::TyFloat(f)) => {
+                                match (i, f) {
+                                    (ast::IntTy::TyI32, ast::FloatTy::TyF32) |
+                                    (ast::IntTy::TyI64, ast::FloatTy::TyF32) |
+                                    (ast::IntTy::TyI64, ast::FloatTy::TyF64) => {
+                                        span_lint(cx, CAST_PRECISION_LOSS, expr.span,
+                                                  &format!("converting from {} to {}, which causes a loss of precision",
+                                                           i, f));
+                                    },
+                                    _ => ()
+                                }
+                            }
+                            (&ty::TypeVariants::TyUint(u), &ty::TypeVariants::TyFloat(f)) => {
+                                match (u, f) {
+                                    (ast::UintTy::TyU32, ast::FloatTy::TyF32) |
+                                    (ast::UintTy::TyU64, ast::FloatTy::TyF32) |
+                                    (ast::UintTy::TyU64, ast::FloatTy::TyF64) => {
+                                        span_lint(cx, CAST_PRECISION_LOSS, expr.span,
+                                                  &format!("converting from {} to {}, which causes a loss of precision",
+                                                           u, f));
+                                    },
+                                    _ => ()
+                                }
+                            },
+                            _ => ()
+                        }
+                    },
+                    (false, true) => {
+                        span_lint(cx, CAST_POSSIBLE_OVERFLOW, expr.span, 
+                                  &format!("the contents of a {} may overflow a {}", cast_from, cast_to));
+                        if !cx.tcx.expr_ty(expr).is_signed() {
+                            span_lint(cx, CAST_SIGN_LOSS, expr.span, 
+                                      &format!("casting from {} to {} loses the sign of the value", cast_from, cast_to));
+                        }
+                    },
+                    (true, true) => {
+                        match (&cast_from.sty, &cast_to.sty) {
+                            (&ty::TypeVariants::TyInt(i1), &ty::TypeVariants::TyInt(i2)) => {
+                                match (i1, i2) {
+                                    (ast::IntTy::TyI64, ast::IntTy::TyI32) |
```

Comment:
What do we do about `TyIs` and `TyUs`? We could treat it like either 32 or 64 bits and warn when either width would cause problems.

I.e. `*64 as *size`, `*size as *32` or smaller, `usize as isize`.